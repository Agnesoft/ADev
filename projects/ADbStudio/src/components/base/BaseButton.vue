<template>
    <button class="btn" @click="handleClick">
        <slot></slot>
    </button>
</template>

<script>
export default {
    name: "BaseButton",
    methods: {
        handleClick(){
            this.$emit('click');
        },
    }
}
</script>

<style>
    .btn{
        cursor: pointer;
        overflow: hidden;
        padding: 10px 20px;
        text-align: center;
        text-decoration: none;
        display: inline-block;
        font-size: 1rem;
        border: none;
        border-radius: 0.2rem;
        --shadow: var(--dark-color) 1px 1px 2px 0px;
        -webkit-box-shadow:  var(--shadow);
        -moz-box-shadow:  var(--shadow);
        box-shadow:  var(--shadow);
    }
    .btn-primary{
        color: var(--light-color);
        background-color: var(--primary-color);
        transition: var(--transition-opacity);
    }
    .btn-primary:hover, .btn-light:hover{
        opacity: 0.7;
    }
    .btn-default{
        border: 1px solid var(--dark-color);
        background-color: var(--light-grey);
        transition: filter 0.4s ease;
    }
    .btn-default:hover{
        filter: brightness(85%)
    }

    .btn-circle{
        border-radius: 50%;
        padding: calc(var(--button-size) * 0.15);
        width: var(--button-size);
        height: var(--button-size);
    }
    .btn-flat{        
        --shadow:  none;   
         border: solid var(--border-width) var(--light-color);
        background-color: var(--primary-color);
        color: var(--light-color);
        padding: calc(var(--button-size) * 0.1);
        font-size: calc(var(--button-size) * 0.6);
        line-height: calc(var(--button-size) * 0.6);
    }
    .btn-light{
        transition: var(--transition-opacity);
        --icon-color: var(--primary-color);
        background: none;
        --shadow:  none; 
    }

</style>
