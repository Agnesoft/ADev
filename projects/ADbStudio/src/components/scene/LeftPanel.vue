<template>

    <SlideXTransition :opened="has_data"  >
        <div class="left-panel">
            <div class="tabs">
                <div class="tab-button" :class="{active: tab.key==active_tab_key}" v-for="tab in tabs" :key="tab.key" :title="tab.title" >
                    <BaseButton class="btn-circle" :class="{'btn-primary' : tab.key==active_tab_key,'btn-light' : tab.key!=active_tab_key}" @click="buttonClicked(tab.key)">
                        <BaseIcon :icon_key="tab.key" />
                    </BaseButton>
                </div>
            </div>
            <SlideXTransition :opened="active_tab_key != -1"  >
                <div class="tab-box">
                    <header class="tab-header">
                        {{tab_title}}
                    </header>
                    <transition name="fade" mode="out-in">
                        <component class="tab-content" v-bind:is="component_content"  v-on="$listeners"></component>
                    </transition>
                </div>
            </SlideXTransition>    
        </div>
    </SlideXTransition> 
</template>

<script>
import SlideXTransition from "../transitions/SlideXTransition";
import { mapGetters } from 'vuex'

export default {
    name: "LeftPanel",
    components: {
        SlideXTransition,
        NodeListData: () => import('./NodeListData'),
        EdgeListData: () => import('./EdgeListData'),
        RawData: () => import('./RawData')
    },
    data() {
        return {
            active_tab_key: -1,
            tabs: [
                {
                    key: 'nodes',
                    title: this.$t('list-of-nodes'),
                    component_content: "NodeListData",
                },
                {
                    key: 'edges',
                    title: this.$t('list-of-edges'),
                    component_content: "EdgeListData",
                },
                {
                    key: 'data',
                    title: this.$t('raw-data'),
                    component_content: "RawData",
                },
            ]
        }
    },
    computed: {
        ...mapGetters({
            data: 'scene/getData',
        }),
        active_tab(){
            if(!this.active_tab_key){
                return false;
            }
            let active_tab_key = this.active_tab_key;
            let tab = this.tabs.find((elem) => elem.key == active_tab_key);
            if(!tab){
                return false;
            }
            return tab;
        },
        component_content(){
            let active_tab_key = this.active_tab_key;
            if(active_tab_key){
                return this.active_tab.component_content;
            }
            return false;
        },
        tab_title(){
            let active_tab_key = this.active_tab_key;
            if(active_tab_key){
                return this.active_tab.title;
            }
            return '';
        },
        has_data(){
            return !(Object.keys(this.data).length === 0 && this.data.constructor === Object);
        }
    },
    methods: {
        buttonClicked(tab_key){
            let tab = this.tabs.find((elem) => elem.key == tab_key);
            if(!tab){
                return;
            }
            if(this.active_tab_key === tab_key){
                this.active_tab_key = -1;
            } else {
                this.active_tab_key = tab_key;
            }
        },
    }
}
</script>

<style scoped>
    .left-panel{
        display: flex;
    }
    .tabs{
        display: flex;
        flex-direction: column;
        margin-right: var(--grid-gap);
    }
    .btn{
        margin: 0.2rem;
        transition: all 0.2s ease;
    }
    .tab-box{
        border-radius: 0.4rem;
        overflow: hidden;
        background: white;
        display: flex;
        flex-direction: column;
        max-height: 80vh;
        --shadow: var(--dark-color) 1px 1px 5px 0px;
        -webkit-box-shadow:  var(--shadow);
        -moz-box-shadow:  var(--shadow);
        box-shadow:  var(--shadow);
    }
    .tab-header{
        padding: 0.5rem;
        white-space: nowrap;
    }
    .tab-content{
        width: 15rem;
        padding: 0.5rem;
        margin: 0;
        height: 100%;
        overflow: auto;
    }
    .fade-enter-active {
        transition: all 0.3s
    }
    .fade-enter {
        opacity: 0;
    }
    .fade-leave-active {
        transition: all 0.3s;
        opacity: 0;
    }
    .tab-header{
        background: var(--primary-color);
        color: var(--light-color);
    }
</style>
