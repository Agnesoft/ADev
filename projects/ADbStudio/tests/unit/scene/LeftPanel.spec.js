import { mount  } from '@vue/test-utils';
import LeftPanel from '@/components/scene/LeftPanel.vue';

import scene_store from '../../../src/store/modules/scene'
import { localVue, i18n,scene_state } from '../TestUtils'
import Vuex from 'vuex';
import BaseButton from "@/components/base/BaseButton.vue";


describe('LeftPanel',() => {
    let wrapper;
    let store;
    let transitionStub; 
    let slideTransitionStub; 

    beforeEach(() => {
        slideTransitionStub = () => ({
            props: {
                opened: Boolean,
            },
            template: '<div v-if="opened" ><slot></slot></div>',
          });
        transitionStub = () => ({
            render: function(h) {
              return this.$options._renderChildren
            }
          });
        localVue.use(Vuex)
        store = new Vuex.Store({
            modules: {
                scene: {
                    namespaced: true,
                    state: scene_state,
                    getters: scene_store.getters
                }
              }
        });
        
        wrapper = mount(LeftPanel, {
            localVue,
            i18n,
            store,
            stubs: {
                transition: transitionStub(),
                SlideXTransition:  slideTransitionStub(),
              }
        });
    })
    it('renders tab buttons',() => {
        let buttons = wrapper.findAll(".tab-button");
        expect(buttons).toHaveLength(3);
        expect(buttons.at(0).attributes('title')).toBe("List of Nodes");
        expect(buttons.at(1).attributes('title')).toBe("List of Edges");
        expect(buttons.at(2).attributes('title')).toBe("Raw Data");
    });

    it("renders tab content after button is clicked",async () => {
        expect(wrapper.find('.tab-box').exists()).toBeFalsy();
        let buttons = wrapper.findAllComponents(BaseButton);
        await buttons.at(0).trigger('click');
        expect(wrapper.find('.tab-box').exists()).toBeTruthy();
    })
    it("changes tab content after another button is clicked", async () => {
        let buttons = wrapper.findAllComponents(BaseButton);
        await buttons.at(0).trigger('click');
        expect(wrapper.find('.tab-header').text()).toBe("List of Nodes");
        await buttons.at(1).trigger('click');
        expect(wrapper.find('.tab-header').text()).toBe("List of Edges");
        await buttons.at(2).trigger('click');
        expect(wrapper.find('.tab-header').text()).toBe("Raw Data");
    })
    it("hides tab content after clicking the same tab button", async () => {
        let buttons = wrapper.findAllComponents(BaseButton);
        await buttons.at(0).trigger('click');
        await buttons.at(0).trigger('click');
        expect(wrapper.find('.tab-box').exists()).toBeFalsy();
    })
})
