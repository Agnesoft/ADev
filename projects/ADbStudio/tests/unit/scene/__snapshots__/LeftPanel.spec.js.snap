// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftPanel renders correctly (snapshot) 1`] = `
<div>
  <div
    class="left-panel"
  >
    <div
      class="tabs"
    >
      <div
        class="tab-button"
        title="List of Nodes"
      >
        <button
          class="btn btn-circle btn-light"
        >
          <svg
            class="base-icon"
            height="100%"
            version="1.1"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              x="0"
              xlink:href="../../assets/images/icons.svg#nodes"
              y="0"
            />
          </svg>
        </button>
      </div>
      <div
        class="tab-button"
        title="List of Edges"
      >
        <button
          class="btn btn-circle btn-light"
        >
          <svg
            class="base-icon"
            height="100%"
            version="1.1"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              x="0"
              xlink:href="../../assets/images/icons.svg#edges"
              y="0"
            />
          </svg>
        </button>
      </div>
      <div
        class="tab-button"
        title="Raw Data"
      >
        <button
          class="btn btn-circle btn-light"
        >
          <svg
            class="base-icon"
            height="100%"
            version="1.1"
            width="100%"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              x="0"
              xlink:href="../../assets/images/icons.svg#data"
              y="0"
            />
          </svg>
        </button>
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;
